### DISCUSSION TITLE
Provide builtin file discovery (events, commands, any pattern)

### DISCUSSION THREAD

**Discussion by pionl** [laravel/framework#44145] [Ideas]
Hi,

With increasing packages and builtin commands (like searching for commands / events) that searching for files in app directory  (for registering its own "map", reading files, etc) it would be great to introduce a way how packages can register its own patterns (or matching logic) and a closure / handle class (with dependency injection in class).

- This single command would make possible to "register" all files for all packages at once
- It could be cache via optimize by Laravel it self
- Registration would be done in ServiceProvider::register

Is this what Laravel would support? I can make the contribution.

Thanks for the info,
Martin

**Comment by Sophist-UK** [comment]
I cannot claim to be an expert about Composer autoload, however it does seem to me that it is not a great idea to autoload masses of files on every single boot, when many of the files loaded may never be called. In a small Laravel application, loading everything might not be a problem, but in a very large application there might be some definite performance overheads for loading unnecessary files.

IMO it would be better for each package to register its own single autoload boot file. Each package can then create an intelligent load strategy that loads the minimal files on actual boot, and then loads additional files when they turn out to be needed. You could even have a default strategy to load every PHP file in `/app/**` which would mimic the current approach.

So for example, a boot to run a command-line call would only need to autoload files needed for the command that is called. Or files used for a particular route could be loaded by middleware depending on the URL route etc.

Composer could provide some callable methods to e.g. load all php files in a specified path-mask e.g. `/modules/*/models/**`, and Laravel might provide some derivatives of this and some middleware functionality.

P.S. I guess that it is not strictly necessary for Composer to change - it is up to each package to define what files are autoloaded by Composer, so I guess that Laravel could create its own bootstrap autoload functionality without any changes to Composer.

**Reply by pionl** [reply]
I get your point, but at this moment Laravel and other packages needs to find the files, for development you do not want to run composer dump-autoload every time you add file / etc, thats why there are optimize commands that is building the map :)

Like this package needs to find the routes https://github.com/spatie/laravel-route-attributes 

By merging how packages / laravel finds files would help speeding the app in dev / production :)

**Reply by Sophist-UK** [reply]
I am not suggesting that you would have to list every file to load individually - simply that the serviceprovider object be changed so that it loads all php files by default but if you override a default method you can then load php files in a specified subset of directories to start with, and then you can define middleware to load specific php files from other directories that you know will be needed for the specific group of routes. In other words, rather than a one-size-fits-all approach that loads all files, we allow developers (if they want to) to have more control over what is loaded and when and with a helper functions to make it easy.

**Reply by pionl** [reply]
Maybe I've incorrectly stated what I mean "load" file - I would provide that file was discovered and you can start building your "cache".

**Reply by Sophist-UK** [reply]
"Cache" is different from "discovery" of php files. 

"Discovery" means recursing the directories and php-importing every php file found - any classes etc. defined in these php file are then made available for all other code to use. This is what the Composer autoload function does.

"Cache" means caching results that take time to create, so that next time you need the same results you can avoid the time-consuming part, and deliver a response to the user faster.

The problem is that if you "discover" every php file, then before you even start to process the web request, php is spending time importing all these files, but many of the classes in these files may never be used. As your app grows in size, the number of files to be imported grows too and php spends and increasing amount of time importing them. But most large apps are modularised - some URLs execute one module, other urls execute a different module - and you probably don't need to load the php files for the 19 modules you are not going to execute this time just so you get the code for the 1 module you are going to execute.

So my idea is that the developer can code in the ServiceProvider which directories / specific php files should be imported always (or if the execution is by a command-line or if you are using Octane) and then middleware is used to recognise which module is being called and import the specific php files for that module.

**Reply by pionl** [reply]
Well i think we see different use cases and probably I'm unable explain it properly in clear english.

Scanning quite large code base already is done by:

- Nova
- Events
- Listeners

If you use modularity and want to use auto-loading you are probably will set "app" instead of "App\Listeners" etc. 

Probably in near future I'll make an experiment and give you real number how "slow" it will be on dev (as mentioned before, "discovery" will not be done on production due the cache file that should be generated).

Thanks for your points

**Reply by Sophist-UK** [reply]
Well either we are talking about completely different things entirely, or one of us doesn't really know what they are talking about. Events and Listeners are an entirely different thing and still involve the php-import of the relevant files. 

Nova is (AFAIK) an admin panel - and that is relevant because you don't need to autoload all the admin panel php files if the URL is not to the admin panel.

**Reply by pionl** [reply]
Check this out https://laravel.com/docs/9.x/events#event-discovery

**Reply by Sophist-UK** [reply]
Thanks for pointing that out. As it says, it uses Reflection i.e. to find classes that have already been created by importing php files. Importing is a pre-requisite for Event Discovery.

**Reply by Sophist-UK** [reply]
Also, Event Discovery is a (lazy? convenient?) alternative to individually listing the listeners in the EventServiceProvider, but it is time consuming so you can cache the discovered list - but you then have to remember to reset the cache, and missing this will lead to bugs.

It feels to me to be better to make the effort to enumerate the listeners in the EventServiceProvider.

**Reply by pionl** [reply]
> Thanks for pointing that out. As it says, it uses Reflection i.e. to find classes that have already been created by importing php files. Importing is a pre-requisite for Event Discovery.

Do not get confused by the terminology, if you check source code it uses Finder class (for filesystem) to find listeners and get the event from class from it and "register" the listener to the event. This is very easy to use and helps you to safe some time (I personally do not use this).

In development you are using this autoload and still it's usable. In production you will call `event:cache` that will cache everything.

What I'm proposing to centralize this:

- 1 "cache" command that would cache results from service providers that uses "file" discovery to register "some stuff" 

Thats all :)

**Anyway just want to point you this:**

If you are importing a class that in PHP file, it does not mean that it is loading the class. It's a just pointer to a class. 

Until you use `new`, `class_implements` and other functions it will try auto load the class (using PHP autoload + composer).

**Reply by Sophist-UK** [reply]
> Do not get confused by the terminology, if you check source code it uses Finder class (for filesystem) to find listeners and get the event from class from it and "register" the listener to the event. This is very easy to use and helps you to safe some time (I personally do not use this).

Do not get confused. Whilst https://github.com/laravel/framework/blob/9.x/src/Illuminate/Foundation/Events/DiscoverEvents.php uses Find, it does **not** repeat **not** import the files, assuming that these files have already been imported. Instead it uses the file name to guess the class used in the file, and then uses Reflection to get an object reference to that class from a string (which only works if the class name matches the file name and if the file has previously been imported. So my comment remains true - that Event discovery only works when the files have already been imported. 

Composer autoload imports php files. Importing the php files reads the source code for the class contained therein and creates the base class in memory i.e. it loads the **class**. (This is NOT an instantiation of the class, but the class definition.) `new` creates an instance of the class, but you can use classes in a static way without instantiating them because they are already loaded.

So what you are talking about is, as I previously said, something entirely different to PHP Compose autoloading or PHP preloading.

**Reply by pionl** [reply]
Dude, I'm not talking about "importing" the files, you are talking about it from first comment.

This discussion is useless due the fact that apparently we are unable to understand each other.  Do not bother responding ;)

Have a good day,

P.S:

Using `use` will not "import" the file... Here is an example https://onlinephp.io/c/ccae4

<img width="1327" alt="Snímek obrazovky 2022-09-20 v 17 36 05" src="https://user-images.githubusercontent.com/1878831/191301861-090412bb-852d-452b-9b93-3108a58773ea.png">

**Reply by Sophist-UK** [reply]
Ok - all that this PHP Sandbox example shows is that the `::class` attribute has a default of returning regardless of whether the class has been imported or not.

TBH I was surprised - I would have expected an exception when it could not find the object to `use`, and another exception when you tried to get the `::class` attribute for a non-existent class. However just because php 8.1 clearly doesn't work as I would expect it to, doesn't prove your point either.

In a recent comment you say

> What I'm proposing to centralize this: 1 "cache" command that would cache results from service providers that uses "file" discovery to register "some stuff"

but I am still completely unclear what you mean by "file discovery" and what "some stuff" actually means. Perhaps some code examples might help clarify this i.e. examples of specific "stuff" that you might want to "discover" and "register" and how you have to do it now and how you might prefer to do it in the future.

Perhaps other people can understand your point as you currently describe it even if I cannot, but if not then I guess that this discussion is not going to go anywhere. 

As for your comment:

> This discussion is useless due the fact that apparently we are unable to understand each other. Do not bother responding ;)

I have said that we seem to be talking about different things and asked you to elaborate on what you are talking about and yet even now it still seems to me that autoload is what you are referring to because you haven't elaborated further except to talk about Event autodiscovery which you mis-described and then used a false example to attempt to illustrate why I am wrong. 

There was no reason for you to be rude. I have been trying to ascertain what it is you are suggesting and telling you how I am interpreting it, but rather than elaborate instead you tell me I don't know how PHP works by referring to Laravel code, introducing that with a snide "Do not get confused." When I check the code and point out that you are incorrect, reflecting your snide remark back at you in doing so, rather than stick to facts and provide a more understandable explanation instead you tell me to shut-up. Please can we keep discussions here both polite and adult in nature.

**Reply by pionl** [reply]
I do not believe I'm being rude. I can say that I feel the same as you do from your style of communication ;) I'm just stating that I'm unable to express my self in way that is clear. I do not have the mental energy to continue this conversation and maybe in future I will be able to provide you better explanation what I mean that will help you understand the meaning of this all. 

I believe this is better for both of use. As I said, have a good day and no hard feelings.

P.S: but thank you for even writing here some response, most of the questions are not answered at all :)
